File: backend/app/services/medical_flow_service.py
"""
Servicio para gestión de flujos médicos normalizados
Proporciona acceso a los 20 flujos médicos completos migrados
"""

from sqlalchemy.orm import Session
from sqlalchemy import text
from typing import List, Optional, Dict, Any
import structlog
from datetime import datetime

logger = structlog.get_logger()

class MedicalFlowService:
    """Servicio para operaciones con flujos médicos normalizados"""
    
    def __init__(self, db: Session):
        self.db = db
    
    def get_specialties(self) -> List[Dict[str, Any]]:
        """Obtener todas las especialidades médicas"""
        try:
            # Query directo a la tabla specialties_normalized
            result = self.db.execute(text("""
                SELECT id, code, name, is_active
                FROM specialties_normalized
                WHERE is_active = 1
                ORDER BY name
            """))
            
            specialties = []
            for row in result:
                specialties.append({
                    "id": row[0],
                    "code": row[1],
                    "name": row[2],
                    "is_active": bool(row[3])
                })
            
            return specialties
            
        except Exception as e:
            logger.error("Error getting specialties", error=str(e))
            return []
    
    def get_flows_by_specialty(self, specialty_id: str) -> List[Dict[str, Any]]:
        """Obtener flujos por especialidad"""
        try:
            # Query directo a la tabla flows
            result = self.db.execute(text("""
                SELECT id, name, specialty_id, description, average_duration, 
                       estimated_cost, is_active
                FROM flows
                WHERE specialty_id = :specialty_id AND is_active = 1
                ORDER BY name
            """), {"specialty_id": specialty_id})
            
            flows = []
            for row in result:
                flows.append({
                    "id": row[0],
                    "name": row[1],
                    "specialty_id": row[2],
                    "description": row[3],
                    "average_duration": row[4],
                    "estimated_cost": float(row[5]) if row[5] else 0.0,
                    "is_active": bool(row[6])
                })
            
            return flows
            
        except Exception as e:
            logger.error("Error getting flows by specialty", error=str(e))
            return []
    
    def search_flows(self, search_term: str) -> List[Dict[str, Any]]:
        """Buscar flujos por término de búsqueda en nombre, descripción o especialidad"""
        try:
            # Query directo a la tabla flows con búsqueda
            search_pattern = f"%{search_term}%"
            result = self.db.execute(text("""
                SELECT f.id, f.name, f.specialty_id, f.description, f.average_duration, 
                       f.estimated_cost, f.is_active,
                       s.name as specialty_name
                FROM flows f
                LEFT JOIN specialties_normalized s ON f.specialty_id = s.id
                WHERE f.is_active = 1
                  AND (f.name LIKE :search OR f.description LIKE :search 
                       OR s.name LIKE :search)
                ORDER BY s.name, f.name
            """), {"search": search_pattern})
            
            flows = []
            for row in result:
                flow_id = row[0]
                flow_data = {
                    "id": flow_id,
                    "name": row[1],
                    "specialty_id": row[2],
                    "description": row[3],
                    "averageDuration": row[4],
                    "estimatedCost": float(row[5]) if row[5] else 0.0,
                    "isActive": bool(row[6]),
                    "specialtyName": row[7] if row[7] else None,
                    "nodes": self.get_flow_nodes(flow_id),
                    "edges": self.get_flow_edges(flow_id)
                }
                flows.append(flow_data)
            
            return flows
            
        except Exception as e:
            logger.error("Error searching flows", error=str(e))
            return []
    
    def get_all_flows(self) -> List[Dict[str, Any]]:
        """Obtener todos los flujos médicos con nodos y edges"""
        try:
            # Query directo a la tabla flows
            result = self.db.execute(text("""
                SELECT f.id, f.name, f.specialty_id, f.description, f.average_duration, 
                       f.estimated_cost, f.is_active,
                       s.name as specialty_name
                FROM flows f
                LEFT JOIN specialties_normalized s ON f.specialty_id = s.id
                WHERE f.is_active = 1
                ORDER BY s.name, f.name
            """))
            
            flows = []
            for row in result:
                flow_id = row[0]
                flow_data = {
                    "id": flow_id,
                    "name": row[1],
                    "specialty_id": row[2],
                    "description": row[3],
                    "averageDuration": row[4],  # Cambiado para coincidir con frontend
                    "estimatedCost": float(row[5]) if row[5] else 0.0,  # Cambiado para coincidir con frontend
                    "isActive": bool(row[6]),  # Cambiado para coincidir con frontend
                    "specialtyName": row[7],  # Cambiado para coincidir con frontend
                    "nodes": self.get_flow_nodes(flow_id),
                    "edges": self.get_flow_edges(flow_id)
                }
                flows.append(flow_data)
            
            return flows
            
        except Exception as e:
            logger.error("Error getting all flows", error=str(e))
            return []
    
    def get_flow_nodes(self, flow_id: str) -> List[Dict[str, Any]]:
        """Obtener nodos de un flujo específico"""
        try:
            # Query directo a la tabla flow_nodes
            result = self.db.execute(text("""
                SELECT id, flow_id, step_type_id, label, description, order_index, 
                       duration_minutes, cost_min, cost_max, cost_avg, position_x, position_y
                FROM flow_nodes
                WHERE flow_id = :flow_id
                ORDER BY order_index
            """), {"flow_id": flow_id})
            
            nodes = []
            for row in result:
                nodes.append({
                    "id": row[0],
                    "flowId": row[1],
                    "stepTypeId": row[2],
                    "label": row[3],
                    "description": row[4],
                    "orderIndex": row[5],
                    "durationMinutes": row[6],
                    "costMin": float(row[7]) if row[7] else 0.0,
                    "costMax": float(row[8]) if row[8] else 0.0,
                    "costAvg": float(row[9]) if row[9] else 0.0,
                    "position": {
                        "x": row[10],
                        "y": row[11]
                    }
                })
            
            return nodes
            
        except Exception as e:
            logger.error("Error getting flow nodes", error=str(e))
            return []
    
    def get_flow_edges(self, flow_id: str) -> List[Dict[str, Any]]:
        """Obtener conexiones de un flujo específico"""
        try:
            # Query directo a la tabla flow_edges_rel
            result = self.db.execute(text("""
                SELECT id, flow_id, source_node_id, target_node_id, edge_type
                FROM flow_edges_rel
                WHERE flow_id = :flow_id
                ORDER BY source_node_id, target_node_id
            """), {"flow_id": flow_id})
            
            edges = []
            for row in result:
                edges.append({
                    "id": row[0],
                    "flowId": row[1],
                    "sourceNodeId": row[2],
                    "targetNodeId": row[3],
                    "edgeType": row[4]
                })
            
            return edges
            
        except Exception as e:
            logger.error("Error getting flow edges", error=str(e))
            return []
    
    def get_flow_by_id(self, flow_id: str) -> Optional[Dict[str, Any]]:
        """Obtener flujo por ID con nodos y conexiones"""
        return self.get_flow_details(flow_id)
    
    def get_flow_details(self, flow_id: str) -> Optional[Dict[str, Any]]:
        """Obtener detalles completos de un flujo"""
        try:
            # Obtener información del flujo
            flow_result = self.db.execute(text("""
                SELECT f.id, f.name, f.specialty_id, f.description, f.average_duration, 
                       f.estimated_cost, f.is_active,
                       s.name as specialty_name
                FROM flows f
                LEFT JOIN specialties_normalized s ON f.specialty_id = s.id
                WHERE f.id = :flow_id
            """), {"flow_id": flow_id})
            
            flow_row = flow_result.fetchone()
            if not flow_row:
                return None
            
            flow_data = {
                "id": flow_row[0],
                "name": flow_row[1],
                "specialty_id": flow_row[2],
                "description": flow_row[3],
                "averageDuration": flow_row[4],  # Cambiado para coincidir con frontend
                "estimatedCost": float(flow_row[5]) if flow_row[5] else 0.0,  # Cambiado para coincidir con frontend
                "isActive": bool(flow_row[6]),  # Cambiado para coincidir con frontend
                "specialtyName": flow_row[7],  # Cambiado para coincidir con frontend
                "nodes": self.get_flow_nodes(flow_id),
                "edges": self.get_flow_edges(flow_id)
            }
            
            return flow_data
            
        except Exception as e:
            logger.error("Error getting flow details", error=str(e))
            return None
    
    def get_referral_criteria(self, specialty_id: str) -> List[Dict[str, Any]]:
        """Obtener criterios de referencia para una especialidad"""
        try:
            # Query directo a la tabla referral_criteria
            result = self.db.execute(text("""
                SELECT id, specialty_id, criteria_type, description, 
                       urgency_level, conditions, metadata
                FROM referral_criteria
                WHERE specialty_id = :specialty_id
                ORDER BY urgency_level, criteria_type
            """), {"specialty_id": specialty_id})
            
            criteria = []
            for row in result:
                criteria.append({
                    "id": row[0],
                    "specialty_id": row[1],
                    "criteria_type": row[2],
                    "description": row[3],
                    "urgency_level": row[4],
                    "conditions": row[5],
                    "metadata": row[6]
                })
            
            return criteria
            
        except Exception as e:
            logger.error("Error getting referral criteria", error=str(e))
            return []
    
    def update_flow(self, flow_id: str, flow_data: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        """Actualizar un flujo médico, incluyendo posiciones de nodos"""
        try:
            # Verificar que el flujo existe
            flow_result = self.db.execute(text("""
                SELECT id FROM flows WHERE id = :flow_id
            """), {"flow_id": flow_id})
            
            if not flow_result.fetchone():
                return None
            
            # Actualizar información del flujo si se proporciona
            if 'name' in flow_data or 'description' in flow_data or 'averageDuration' in flow_data or 'estimatedCost' in flow_data:
                update_fields = []
                params = {"flow_id": flow_id}
                
                if 'name' in flow_data:
                    update_fields.append("name = :name")
                    params["name"] = flow_data['name']
                if 'description' in flow_data:
                    update_fields.append("description = :description")
                    params["description"] = flow_data['description']
                if 'averageDuration' in flow_data:
                    update_fields.append("average_duration = :average_duration")
                    params["average_duration"] = flow_data['averageDuration']
                if 'estimatedCost' in flow_data:
                    update_fields.append("estimated_cost = :estimated_cost")
                    params["estimated_cost"] = flow_data['estimatedCost']
                
                if update_fields:
                    update_query = f"""
                        UPDATE flows 
                        SET {', '.join(update_fields)}
                        WHERE id = :flow_id
                    """
                    self.db.execute(text(update_query), params)
                    self.db.commit()
            
            # Actualizar posiciones de nodos si se proporcionan
            if 'nodes' in flow_data and isinstance(flow_data['nodes'], list):
                for node_data in flow_data['nodes']:
                    if 'id' in node_data and 'position' in node_data:
                        position = node_data['position']
                        if isinstance(position, dict) and 'x' in position and 'y' in position:
                            self.db.execute(text("""
                                UPDATE flow_nodes 
                                SET position_x = :x, position_y = :y
                                WHERE id = :node_id AND flow_id = :flow_id
                            """), {
                                "x": int(position['x']),
                                "y": int(position['y']),
                                "node_id": node_data['id'],
                                "flow_id": flow_id
                            })
                
                self.db.commit()
            
            # Retornar el flujo actualizado
            return self.get_flow_by_id(flow_id)
            
        except Exception as e:
            logger.error("Error updating flow", error=str(e))
            self.db.rollback()
            return None
    
    def get_step_types(self) -> List[Dict[str, Any]]:
        """Obtener todos los tipos de pasos"""
        try:
            result = self.db.execute(text("""
                SELECT id, code, name
                FROM step_types
                ORDER BY name
            """))
            
            step_types = []
            for row in result:
                step_types.append({
                    "id": row[0],
                    "code": row[1],
                    "name": row[2]
                })
            
            return step_types
        except Exception as e:
            logger.error("Error getting step types", error=str(e))
            return []
    
    def get_urgency_levels(self) -> List[Dict[str, Any]]:
        """Obtener todos los niveles de urgencia"""
        try:
            result = self.db.execute(text("""
                SELECT id, code, name
                FROM urgency_levels
                ORDER BY name
            """))
            
            urgency_levels = []
            for row in result:
                urgency_levels.append({
                    "id": row[0],
                    "code": row[1],
                    "name": row[2]
                })
            
            return urgency_levels
        except Exception as e:
            logger.error("Error getting urgency levels", error=str(e))
            return []
    
    def get_referral_criteria_by_specialty(self, specialty_id: str) -> List[Dict[str, Any]]:
        """Obtener criterios de referencia por especialidad"""
        return self.get_referral_criteria()
    
    def get_flow_statistics(self) -> Dict[str, Any]:
        """Obtener estadísticas de flujos"""
        try:
            flows = self.get_all_flows()
            total_nodes = sum(len(flow.get('nodes', [])) for flow in flows)
            total_edges = sum(len(flow.get('edges', [])) for flow in flows)
            avg_duration = sum(flow.get('averageDuration', 0) for flow in flows) / len(flows) if flows else 0
            avg_cost = sum(flow.get('estimatedCost', 0) for flow in flows) / len(flows) if flows else 0
            
            return {
                "totalFlows": len(flows),
                "totalNodes": total_nodes,
                "totalEdges": total_edges,
                "averageDuration": avg_duration,
                "averageCost": avg_cost
            }
        except Exception as e:
            logger.error("Error getting flow statistics", error=str(e))
            return {}

File: client/src/services/medicalFlowService.ts
/**
 * Servicio frontend para flujos médicos normalizados
 * Proporciona acceso a los 20 flujos médicos completos con datos reales
 */

import axios from 'axios';

const API_URL = import.meta.env.VITE_API_URL || 'http://localhost:8000';

export interface Specialty {
  id: string;
  code: string;
  name: string;
  isActive: boolean;
}

export interface StepType {
  id: string;
  code: string;
  name: string;
}

export interface UrgencyLevel {
  id: string;
  code: string;
  name: string;
}

export interface FlowNode {
  id: string;
  flowId: string;
  stepTypeId: string;
  stepTypeName?: string;
  label: string;
  description?: string;
  orderIndex: number;
  durationMinutes?: number;
  costMin?: number;
  costMax?: number;
  costAvg?: number;
  position: {
    x: number;
    y: number;
  };
  createdAt: string;
}

export interface FlowEdge {
  id: string;
  flowId: string;
  sourceNodeId: string;
  targetNodeId: string;
  edgeType: string;
  createdAt: string;
}

export interface Flow {
  id: string;
  name: string;
  specialtyId?: string;
  specialtyName?: string;
  description?: string;
  averageDuration?: number;
  estimatedCost?: number;
  isActive: boolean;
  createdAt?: string;
  updatedAt?: string;
  nodes: FlowNode[];
  edges: FlowEdge[];
}

export interface ReferralCriteria {
  id: string;
  diagnosis: string;
  criterion: string;
  targetSpecialtyId: string;
  targetSpecialtyName?: string;
  urgencyId: string;
  urgencyName?: string;
  isActive: boolean;
  createdAt: string;
}

export interface FlowStatistics {
  totalSpecialties: number;
  totalFlows: number;
  totalNodes: number;
  totalEdges: number;
  totalReferralCriteria: number;
  averageFlowDuration: number;
  averageFlowCost: number;
}

export interface FlowFilters {
  specialtyId?: string;
  minDuration?: number;
  maxDuration?: number;
  minCost?: number;
  maxCost?: number;
  search?: string;
}

class MedicalFlowService {
  private baseURL = `${API_URL}/api/v1/medical-flows`;

  /**
   * Obtener todas las especialidades médicas
   */
  async getSpecialties(): Promise<Specialty[]> {
    try {
      const response = await axios.get(`${this.baseURL}/specialties`);
      return response.data.data || [];
    } catch (error) {
      console.error('Error obteniendo especialidades:', error);
      return [];
    }
  }

  /**
   * Obtener especialidad por ID
   */
  async getSpecialtyById(specialtyId: string): Promise<Specialty | null> {
    try {
      const response = await axios.get(`${this.baseURL}/specialties/${specialtyId}`);
      return response.data.data || null;
    } catch (error) {
      console.error(`Error obteniendo especialidad ${specialtyId}:`, error);
      return null;
    }
  }

  /**
   * Obtener todos los flujos médicos
   */
  async getFlows(filters?: FlowFilters): Promise<Flow[]> {
    try {
      const params = new URLSearchParams();
      
      if (filters?.specialtyId) params.append('specialty_id', filters.specialtyId);
      if (filters?.minDuration) params.append('min_duration', filters.minDuration.toString());
      if (filters?.maxDuration) params.append('max_duration', filters.maxDuration.toString());
      if (filters?.minCost) params.append('min_cost', filters.minCost.toString());
      if (filters?.maxCost) params.append('max_cost', filters.maxCost.toString());
      if (filters?.search) params.append('search', filters.search);

      const queryString = params.toString();
      const url = queryString ? `${this.baseURL}/flows?${queryString}` : `${this.baseURL}/flows`;
      
      const response = await axios.get(url);
      return response.data.data || [];
    } catch (error) {
      console.error('Error obteniendo flujos:', error);
      return [];
    }
  }

  /**
   * Obtener flujos por especialidad
   */
  async getFlowsBySpecialty(specialtyId: string): Promise<Flow[]> {
    return this.getFlows({ specialtyId });
  }

  /**
   * Obtener flujo completo por ID
   */
  async getFlowById(flowId: string): Promise<Flow | null> {
    try {
      const response = await axios.get(`${this.baseURL}/flows/${flowId}`);
      return response.data.data || null;
    } catch (error) {
      console.error(`Error obteniendo flujo ${flowId}:`, error);
      return null;
    }
  }

  /**
   * Obtener tipos de pasos
   */
  async getStepTypes(): Promise<StepType[]> {
    try {
      const response = await axios.get(`${this.baseURL}/step-types`);
      return response.data.data || [];
    } catch (error) {
      console.error('Error obteniendo tipos de pasos:', error);
      return [];
    }
  }

  /**
   * Obtener niveles de urgencia
   */
  async getUrgencyLevels(): Promise<UrgencyLevel[]> {
    try {
      const response = await axios.get(`${this.baseURL}/urgency-levels`);
      return response.data.data || [];
    } catch (error) {
      console.error('Error obteniendo niveles de urgencia:', error);
      return [];
    }
  }

  /**
   * Obtener criterios de referencia
   */
  async getReferralCriteria(specialtyId?: string): Promise<ReferralCriteria[]> {
    try {
      const params = specialtyId ? `?specialty_id=${specialtyId}` : '';
      const response = await axios.get(`${this.baseURL}/referral-criteria${params}`);
      return response.data.data || [];
    } catch (error) {
      console.error('Error obteniendo criterios de referencia:', error);
      return [];
    }
  }

  /**
   * Obtener estadísticas de flujos
   */
  async getFlowStatistics(): Promise<FlowStatistics | null> {
    try {
      const response = await axios.get(`${this.baseURL}/statistics`);
      return response.data.data || null;
    } catch (error) {
      console.error('Error obteniendo estadísticas:', error);
      return null;
    }
  }

  /**
   * Buscar flujos por texto
   */
  async searchFlows(query: string): Promise<Flow[]> {
    return this.getFlows({ search: query });
  }

  /**
   * Obtener flujos por rango de duración
   */
  async getFlowsByDurationRange(minDuration: number, maxDuration: number): Promise<Flow[]> {
    return this.getFlows({ minDuration, maxDuration });
  }

  /**
   * Obtener flujos por rango de costo
   */
  async getFlowsByCostRange(minCost: number, maxCost: number): Promise<Flow[]> {
    return this.getFlows({ minCost, maxCost });
  }

  /**
   * Formatear duración en minutos a texto legible
   */
  formatDuration(minutes: number): string {
    if (minutes < 60) {
      return `${minutes} min`;
    }
    const hours = Math.floor(minutes / 60);
    const remainingMinutes = minutes % 60;
    return remainingMinutes > 0 ? `${hours}h ${remainingMinutes}min` : `${hours}h`;
  }

  /**
   * Formatear costo a moneda
   */
  formatCost(cost: number): string {
    return new Intl.NumberFormat('es-CO', {
      style: 'currency',
      currency: 'COP',
      minimumFractionDigits: 0,
      maximumFractionDigits: 0
    }).format(cost);
  }

  /**
   * Obtener color para tipo de paso
   */
  getStepTypeColor(stepTypeCode: string): string {
    const colors: { [key: string]: string } = {
      // Códigos del backend
      'st-cons': '#2196F3',      // Azul - Consulta
      'st-lab': '#4CAF50',       // Verde - Laboratorio
      'st-img': '#FF9800',       // Naranja - Imagenología
      'st-dx': '#9C27B0',        // Púrpura - Diagnóstico
      'st-rx': '#F44336',        // Rojo - Tratamiento/Prescripción
      'st-ref': '#607D8B',       // Gris azulado - Referencia
      'st-fu': '#795548',        // Marrón - Seguimiento
      'st-proc': '#3F51B5',      // Índigo - Procedimiento
      'st-emergency': '#E91E63', // Rosa - Emergencia
      'st-discharge': '#009688', // Teal - Alta
      
      // Códigos legacy (mantener compatibilidad)
      'consultation': '#2196F3',
      'laboratory': '#4CAF50',
      'imaging': '#FF9800',
      'diagnosis': '#9C27B0',
      'prescription': '#F44336',
      'referral': '#607D8B',
      'followup': '#795548',
      'emergency': '#E91E63',
      'procedure': '#3F51B5',
      'discharge': '#009688'
    };
    return colors[stepTypeCode] || '#757575';
  }

  /**
   * Obtener icono para tipo de paso
   */
  getStepTypeIcon(stepTypeCode: string): string {
    const icons: { [key: string]: string } = {
      // Códigos del backend
      'st-cons': 'medical_services',    // Consulta
      'st-lab': 'science',              // Laboratorio
      'st-img': 'camera_alt',           // Imagenología
      'st-dx': 'psychology',            // Diagnóstico
      'st-rx': 'medication',            // Tratamiento/Prescripción
      'st-ref': 'arrow_forward',        // Referencia
      'st-fu': 'schedule',              // Seguimiento
      'st-proc': 'medical_information', // Procedimiento
      'st-emergency': 'emergency',      // Emergencia
      'st-discharge': 'exit_to_app',    // Alta
      
      // Códigos legacy (mantener compatibilidad)
      'consultation': 'medical_services',
      'laboratory': 'science',
      'imaging': 'camera_alt',
      'diagnosis': 'psychology',
      'prescription': 'medication',
      'referral': 'arrow_forward',
      'followup': 'schedule',
      'emergency': 'emergency',
      'procedure': 'medical_information',
      'discharge': 'exit_to_app'
    };
    return icons[stepTypeCode] || 'help';
  }
}

export const medicalFlowService = new MedicalFlowService();
export default medicalFlowService;

